File: leetcode/dfs bfs/127. Word Ladder/js/main.js
// BFS
// Time O(M * N) где M - длина слова, а N - общее количество слов в списке входных слов.
// Space O(M * N)
const ladderLength = (beginWord, endWord, wordList) => {
  let queue = [beginWord];
  let visited = Array(wordList.length).fill(false);
  let cnt = 1;

  while (queue.length) {
    let size = queue.length;

    for (let i = 0; i < size; i++) {
      let s1 = queue.shift();

      if (s1 == endWord) {
        return cnt;
      }

      for (let i = 0; i < wordList.length; i++) {
        if (visited[i] || !oneEditReplace(s1, wordList[i])) continue;

        visited[i] = true;
        queue.push(wordList[i]);
      }
    }
    cnt++;
  }

  return 0;
};

function oneEditReplace(s1, s2) {
  if (s1 === s2) {
    return false;
  }

  let foundDifference = false;

  for (let i = 0; i < s1.length; i++) {
    if (s1[i] !== s2[i]) {
      if (foundDifference) {
        return false;
      }

      foundDifference = true;
    }
  }

  return true;
}

File: leetcode/dfs bfs/127. Word Ladder/py/main.py
import collections

class Solution:
    # Time O(M * N) где M - длина слова, а N - общее количество слов в списке входных слов.
    # Space O(M * N)
    def ladderLength(self, beginWord, endWord, wordList):
        all_combo_dict = collections.defaultdict(list)
        queue = collections.deque([(beginWord, 1)])
        visited = set([beginWord])
        
        all_combo_dict = collections.defaultdict(list)
        
        for word in wordList:
            for i in range(len(word)):
                all_combo_dict[word[:i] + "*" + word[i+1:]].append(word)

        while queue:
          size = len(queue)

          for _ in range(size):
            current, level = queue.popleft()

            for i in range(len(current)):
              intermediate_word = current[:i] + '*' + current[i+1:]

              for word in all_combo_dict[intermediate_word]:
                if word == endWord:
                  return level + 1
                
                if word not in visited:
                  visited.add(word)
                  queue.append([word, level + 1])

              all_combo_dict[intermediate_word] = []
          
        return 0